### PR TITLE
Add tools usage option to kiosk page

### PR DESCRIPTION
**Pull request by r-xyz**
Hello,

I open this PR (still work in progress) to check if the feature makes sense and to get some feedback on the best way to implement it.

The feature would be beneficial for labs which do not have a jumbotron. The idea is to add additional options to the kiosk so it could show:

- Tools currently in use
- Alerts

It is something we currently achieved with a ugly hack which basically pulls data from the Jumbotron into the kiosk page.
After noticing a [similar feature](https://github.com/usnistgov/NEMO/commit/8c89716065f87706502be230bc811729f93545c4) had been implemented for area occupancy, I though it might be worth to extend the same model to usage and Alerts.

After reconsidering, I tried to pull data from the status dashboard (since it allows to discriminate between categories).
image

### CODE REVIEW

**Comment by r-xyz** [comment]
Added filtering of tools.
Should I add the color legend also to the Status Dashboard/Tools tab, for coherence with the Occupancy tab?
image
Right now  it is not present:

image

**Comment by rptmat57** [comment]
Yes 100% for adding the color legend in the tool status tab.

**Comment by r-xyz** [comment]
The basic functionality is ready and functioning. 

- setting parameter `usage`(e.g. `=1`), enables the tools status section
- setting parameter `alerts`(e.g. `=1`), enables the alerts section

image
The extra section only appears in the kiosk welcome page (`Scan your badge to control tools`)

The PR should be ok, but please double check.

PS: I have an extra suggestion which I could try to address in a separate PR, if accepted.
Would it make sense to make the code for Jumbotron, Status Dashboard (and this new functionality) modular? This could reduce maintenance effort, since the data shown is mostly the same.
I would probably centralize the function to the status dashboard, and have it called by the Jumbotron.
Similarly, alerts section could also be moved to an external html template, which can be called both by `landing.html` page as well as `kiosk.html`.
Just a proposal; I am still partially familiar with the project codebase, so I might be missing some pitfalls.

**Comment by rptmat57** [comment]
Hi, I think it would be a great idea to make the refactor the code for jumbotron, status dashboard and this to use similar views/templates, as long as it doesn't overcomplicated things.

File: NEMO/apps/kiosk/templates/kiosk/kiosk.html [hunk 1]
```diff
@@ -322,6 +342,31 @@ <h3 id="interlock_message" class="text-center"></h3>
...
+                node.nodeValue = css_text;
+                return node;
+            })(css_text);
+        }
+
+        function fetch_usage()
+		{
+            $("#usage_title").html("<h2>Tools in use</h2>")
+			ajax_get('{% url 'kiosk_usage' %}', {'interest': 'tools'}, function(response) { $("#usage").html(response); });
+            set_style('.in_use { display: table-row; } .idle { display: none; }');
```
Comment (rptmat57) [inline NEMO/apps/kiosk/templates/kiosk/kiosk.html:364]: I might be missing something here but wouldn't be easier to use jquery in the callback `function(response)` to show/hide stuff?
for example `$(".in_use").css("display", "table-row");` and `$(".idle").hide()`

**Review by rptmat57 — COMMENTED** [review]
Looks good, thank you!
Check out my comment about the `set_style` piece

File: NEMO/apps/kiosk/templates/kiosk/kiosk.html (hunk 1 above)
Comment (r-xyz) [reply NEMO/apps/kiosk/templates/kiosk/kiosk.html]: Good point, thanks, I changed it!

File: NEMO/apps/kiosk/templates/kiosk/kiosk.html [hunk 2]
```diff
@@ -322,6 +342,36 @@ <h3 id="interlock_message" class="text-center"></h3>
...
+            sheet.type = 'text/css';
+            (document.head || document.getElementsByTagName('head')[0]).appendChild(sheet);
+            return (set_style = function(css_text, node)
+            {
+                if(!node || node.parentNode !== sheet)
+                    return sheet.appendChild(document.createTextNode(css_text));
+                node.nodeValue = css_text;
+                return node;
+            })(css_text);
+        }
```
Comment (rptmat57) [inline NEMO/apps/kiosk/templates/kiosk/kiosk.html:358]: this can now be removed
```suggestion
```

**Comment by rptmat57** [comment]
Yes, the refactoring should be in a separate PR.

Also, for consistency, this should also be added to the `apps.area_access` subplugin (same as the occupancy pulling)

I'll let you decide if you prefer to add it to this PR or if it's easier to do in the refactoring PR.

Thanks, this is great!

File: NEMO/apps/kiosk/templates/kiosk/kiosk.html (hunk 2 above)
Comment (r-xyz) [reply NEMO/apps/kiosk/templates/kiosk/kiosk.html]: Fixed by [f10072d](https://github.com/usnistgov/NEMO/pull/178/commits/f10072d3cf4bcdfd12e4a80726136d00d577decf)

File: NEMO/apps/kiosk/templates/kiosk/alerts.html [hunk 3]
```diff
@@ -0,0 +1,32 @@
+{% if alerts or disabled_resources %}
+    <div class="{% if areas_exist and tools_exist %}col-lg-4{% else %}col-lg-6{% endif %}" style="text-align:center">
```
[the remark this answers is not in the document]
Comment (r-xyz) [reply NEMO/apps/kiosk/templates/kiosk/alerts.html]: My bad, I took it from the jumbotron but forgot to fix this part.
Fixed by [45f502a](https://github.com/usnistgov/NEMO/pull/178/commits/45f502af116e743be36c9efdbd6fb0ad2b608b41).

**Comment by r-xyz** [comment]
I already moved alerts under `apps.area_access` so there is less to change for next PR.
I will start working on it hopefully next week, but this one should be completed (unless I missed something else).
Thanks

**Comment by r-xyz** [comment]
image

**Comment by r-xyz** [comment]
Changed very last detail.
image

**Comment by rptmat57** [comment]
Thanks, after a bit more reflection on occupancy and alerts etc. I believe it makes more sense to have all of those partial views somewhere in NEMO itself. They could be in a `partial_views.py` or something like that and have the templates in the `snippets` template folder.
Since they will be used by the jumbotron which is in NEMO, it seems more logical to have them in NEMO.
Nothing to do here, but something to keep in mind for the refactoring

**Comment by rptmat57** [comment]
Also if at all possible, I would prefer just checking the presence of the `alerts` and `usage` parameters rather than if they have a value. i.e. `nemo.nist.gov/kiosk?alerts&usage`

Great work, thanks a lot!

**Comment by rptmat57** [comment]
Hi @r-xyz, I know this one was closed a while back but any chance the same parameters could be added to `welcome_screen` and `farewell_screen` for consistency?

**Pull request merged**